File: days/day_12/main.py
import time
import os

compass = {
	'N' : [0,1],
	'E' : [1,0],
	'S' : [0,-1],
	'W' : [-1,0]
}


def add(u, v):
	return [u[i]+v[i] for i in range(len(u))]


def mult(k, u):
	return [k*u[i] for i in range(len(u))]


def rotate_left(u, deg):
	x, y = u
	while deg < 0:
		deg += 360

	if deg == 90:
		return [-y,x]
	elif deg == 180:
		return [-x,-y]
	elif deg == 270:
		return [y, -x]
	else:
		return [x, y]


def part_one(instructions):
	direction = [1,0]
	pos = [0,0]

	for i in range(len(instructions)):
		letter, num = instructions[i]

		if letter in compass:
			pos = add(pos, mult(num, compass[letter]))
		elif letter == 'F':
			pos = add(pos, mult(num, direction))
		else:
			deg = num

			if letter == 'R':
				deg *= -1

			direction = rotate_left(direction, deg)

	return sum(abs(x) for x in pos) 


def part_two(instructions):
	pos = [0,0]
	waypoint = [10,1]

	for i in range(len(instructions)):
		letter, num = instructions[i]

		if letter in compass:
			waypoint = add(waypoint, mult(num, compass[letter]))
		elif letter == 'F':
			pos = add(pos, mult(num, waypoint))
		else:
			deg = num

			if letter == 'R':
				deg *= -1

			waypoint = rotate_left(waypoint, deg)

	return sum(abs(x) for x in pos)


def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		instructions = [[line[0],int(line[1:])] for line in data.readlines()]
		
		part_one_ans = part_one(instructions)
		part_two_ans = part_two(instructions)

		print('Day 12 ({:,.3f}s)'.format(time.time()-start_time))
		print('  Part 1: {}'.format(part_one_ans))
		print('  Part 2: {}'.format(part_two_ans))


if __name__ == "__main__":
		main()

File: days/day_12/main.js
const fs = require('fs');
const readline = require('readline');

const compass = {
	'N' : [0,1],
	'E' : [1,0],
	'S' : [0,-1],
	'W' : [-1,0]
}
const directions = Object.keys(compass)

async function loadInput() {
	const rl = readline.createInterface({
		input: fs.createReadStream(__dirname + '/input.txt')
	});

	const lines = [];
	for await (const line of rl) {
		lines.push(line);
	}
	return lines
}

function add(u, v) {
	return u.map((u_i, i) => u_i + v[i])
}

function mult(k, u) {
	return u.map(u_i => k * u_i)
}

function rotateCCW(u, deg) {
	const [x, y] = u;
	
	while (deg < 0) {
		deg += 360;
	}

	switch (deg) {
		case 90:
			return [-y,x];
		case 180:
			return [-x,-y];
		case 270:
			return [y, -x];
		default:
			return [x, y];
	}

}


function partOne(instructions) {
	let direction = [1,0]
	let pos = [0,0]

	for (const [letter, num] of instructions) {
		if (directions.includes(letter)) {
			pos = add(pos, mult(num, compass[letter]))
		}
		else if (letter == 'F') {
			pos = add(pos, mult(num, direction))
		}
		else {
			deg = num

			if (letter == 'R') {
				deg *= -1
			}

			direction = rotateCCW(direction, deg)
		}
	}

	return pos.reduce((total, p) => total + Math.abs(p), 0);
}

function partTwo(instructions) {
	let pos = [0,0]
	let waypoint = [10,1]

	for (const [letter, num] of instructions) {
		if (directions.includes(letter)) {
			waypoint = add(waypoint, mult(num, compass[letter]))
		}
		else if (letter == 'F') {
			pos = add(pos, mult(num, waypoint))
		}
		else {
			deg = num

			if (letter == 'R') {
				deg *= -1
			}

			waypoint = rotateCCW(waypoint, deg)
		}
	}

	return pos.reduce((total, p) => total + Math.abs(p), 0);
}


async function main() {
	const start_time = process.hrtime() 
	const lines = await loadInput();

	const instructions = lines.map(line => [line[0],Number(line.substring(1))])

	const part_one_ans = partOne(instructions)
	const part_two_ans = partTwo(instructions)
	
	const diff = process.hrtime(start_time)
	const total_time = (diff[0] + diff[1]/1e9).toFixed(3)
	
	console.log(`Day 12 (${total_time}s)`)
	console.log(`  Part 1: ${part_one_ans}`);
	console.log(`  Part 2: ${part_two_ans}`);
}

if (require.main === module) {
	main();
}